src/meta/src/stream/stream_graph/schedule.rs

```diff
@@ -290,22 +281,55 @@ impl Scheduler {
 
         let mut facts = Vec::new();
 
-        // Building fragments and Singletons
+        // Singletons.
         for (&id, fragment) in graph.building_fragments() {
-            facts.push(Fact::Fragment(id));
             if fragment.requires_singleton {
-                facts.push(Fact::SingletonReq(id));
+                facts.push(Fact::Req {
+                    id,
+                    req: Req::AnySingleton,
+                });
             }
         }
-        // External
-        for (id, req) in existing_distribution {
-            let dist = match req {
-                Distribution::Singleton(worker_slot_id) => DistId::Singleton(worker_slot_id),
-                Distribution::Hash(mapping) => DistId::Hash(hash_mapping_id[&mapping]),
+        // Vnode count requirements: if a fragment is going to look up an existing table,
+        // it must have the same vnode count as that table.
+        for (&id, fragment) in graph.building_fragments() {
+            visit_fragment(&mut (*fragment).clone(), |node| {
+                use risingwave_pb::stream_plan::stream_node::NodeBody;
+                let vnode_count = match node {
+                    NodeBody::StreamScan(node) => {
+                        if let Some(table) = &node.arrangement_table {
+                            table.vnode_count()
+                        } else if let Some(table) = &node.table_desc {
+                            table.vnode_count()
+                        } else {
+                            return;
+                        }
+                    }
+                    NodeBody::TemporalJoin(node) => node.get_table_desc().unwrap().vnode_count(),
```

Comment:
May not be required as it has no-shuffle exchange from the `UpstreamOnly` chain, then the upstream table. Still add here for consistency and as an assertion.